java/feathub-udf/flink-udf/src/main/java/com/alibaba/feathub/flink/udf/processfunction/SlidingWindowKeyedProcessFunction.java

```diff
@@ -96,93 +142,211 @@ public void onTimer(
+        final int timestampListSize = timestampList.size();
+
+        Row outputRow = Row.withNames();
+        for (int i = 0; i < keyFieldNames.length; i++) {
+            outputRow.setField(keyFieldNames[i], ctx.getCurrentKey().getField(i));
+        }
+
+        Row rowToAdd = state.timestampToRow.get(timestamp);
+        final Row accumulators = state.getAccumulators();
+        final Map<String, Integer> leftIdxMap = state.getLeftTimestampIdx();
+
+        for (AggregationFieldsDescriptor.AggregationFieldDescriptor descriptor :
+                aggregationFieldsDescriptor.getAggFieldDescriptors()) {
+            Object accumulator = accumulators.getField(descriptor.outFieldName);
+            if (accumulator == null) {
+                accumulator = descriptor.aggFunc.createAccumulator();
+                accumulators.setField(descriptor.outFieldName, accumulator);
+            }
+            if (rowToAdd != null) {
+                descriptor.aggFunc.add(
+                        accumulator, rowToAdd.getField(descriptor.inFieldName), timestamp);
             }
-            Row curRow = state.timestampToRow.get(rowTime);
 
-            // TODO: Optimize by supporting retract value from aggregation function.
-            for (AggregationFieldsDescriptor.AggregationFieldDescriptor descriptor :
-                    aggregationFieldsDescriptor.getAggFieldDescriptors()) {
-                if (timestamp - descriptor.windowSizeMs >= rowTime) {
-                    continue;
+            // Advance left idx and retract values
+            int leftIdx = leftIdxMap.getOrDefault(descriptor.outFieldName, 0);
+            for (; leftIdx < timestampListSize; ++leftIdx) {
+                long rowTime = timestampList.get(leftIdx);
+                if (timestamp - descriptor.windowSizeMs < rowTime) {
+                    if (rowTime <= timestamp) {
+                        hasRow = true;
+                    }
+                    break;
                 }
-                descriptor.aggFunc.aggregate(curRow.getField(descriptor.inFieldName), rowTime);
-                hasRow = true;
+                Row curRow = state.timestampToRow.get(rowTime);
+                descriptor.aggFunc.retract(accumulator, curRow.getField(descriptor.inFieldName));
             }
+            leftIdxMap.put(descriptor.outFieldName, leftIdx);
+
+            outputRow.setField(descriptor.outFieldName, descriptor.aggFunc.getResult(accumulator));
         }
+        state.updateLeftTimestampIdx(leftIdxMap);
+        state.updateAccumulators(accumulators);
+        outputRow.setField(rowTimeFieldName, Instant.ofEpochMilli(timestamp));
 
-        state.pruneRow(
-                ctx.getCurrentKey(), timestamp - aggregationFieldsDescriptor.getMaxWindowSizeMs());
+        state.pruneRow(timestamp, aggregationFieldsDescriptor);
 
         if (!hasRow) {
-            // output nothing if no row is aggregated.
+            state.maxRegisteredTimer.clear();
```

Comment:
Is the goal here to clean up memory usage for keys who do not have any non-expired rows? If so, would it be simpler to move this logic to `pruneRow()`so that `pruneRow()` consistently handles state clearing for all states defined in `SlidingWindowState`?

Same for `state.lastOutputRow.clear()`.